#!/usr/bin/env python3
"""
gRPC Contract Tests - PR-6
–ê–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏–µ —Ç–µ—Å—Ç—ã –Ω–∞ –æ—Å–Ω–æ–≤–µ –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–∞–±–ª–∏—Ü –∏–∑ GRPC_PROTOCOL_AUDIT.md

–ü—Ä–æ–≤–µ—Ä—è–µ—Ç 8-12 —Å—Ü–µ–Ω–∞—Ä–∏–µ–≤ –∏–∑ –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–∞–±–ª–∏—Ü:
- –í–∞–ª–∏–¥–Ω—ã–µ/–Ω–µ–≤–∞–ª–∏–¥–Ω—ã–µ –≤—Ö–æ–¥—ã
- –¢–∞–π–º–∞—É—Ç—ã
- –û–±—Ä—ã–≤—ã —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è
- –ö–æ–¥—ã –æ—à–∏–±–æ–∫
"""

import sys
import os
import asyncio
import logging
from pathlib import Path
from typing import Dict, Any, List, Optional
from dataclasses import dataclass
from enum import Enum

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ –º–æ–¥—É–ª—è–º
project_root = Path(__file__).parent.parent
sys.path.insert(0, str(project_root))

try:
    import grpc
    from grpc import aio
    
    # –ò–º–ø–æ—Ä—Ç protobuf —Ñ–∞–π–ª–æ–≤
    sys.path.insert(0, str(project_root / "modules" / "grpc_service"))
    import streaming_pb2
    import streaming_pb2_grpc
except ImportError as e:
    print(f"‚ùå –û—à–∏–±–∫–∞ –∏–º–ø–æ—Ä—Ç–∞: {e}")
    print("–£–±–µ–¥–∏—Ç–µ—Å—å, —á—Ç–æ —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω—ã –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏: pip install grpcio grpcio-tools")
    sys.exit(1)

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)


class TestResult(Enum):
    """–†–µ–∑—É–ª—å—Ç–∞—Ç —Ç–µ—Å—Ç–∞"""
    PASS = "PASS"
    FAIL = "FAIL"
    SKIP = "SKIP"


@dataclass
class ContractTest:
    """–ö–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–µ—Å—Ç"""
    name: str
    method: str
    input_data: Dict[str, Any]
    expected_output: Optional[Dict[str, Any]] = None
    expected_error_code: Optional[str] = None
    expected_latency_ms: Optional[float] = None
    should_fail: bool = False
    description: str = ""


class GrpcContractTester:
    """–¢–µ—Å—Ç–µ—Ä –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–∞–±–ª–∏—Ü"""
    
    def __init__(self, host: str, port: int):
        """
        –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è —Ç–µ—Å—Ç–µ—Ä–∞
        
        Args:
            host: –•–æ—Å—Ç —Å–µ—Ä–≤–µ—Ä–∞
            port: –ü–æ—Ä—Ç —Å–µ—Ä–≤–µ—Ä–∞
        """
        self.host = host
        self.port = port
        self.channel = None
        self.stub = None
        
        # –†–µ–∑—É–ª—å—Ç–∞—Ç—ã —Ç–µ—Å—Ç–æ–≤
        self.results: List[tuple[ContractTest, TestResult, str]] = []
    
    async def connect(self) -> bool:
        """–ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ —Å–µ—Ä–≤–µ—Ä—É"""
        try:
            address = f"{self.host}:{self.port}"
            
            if self.port == 443:
                credentials = grpc.ssl_channel_credentials()
                self.channel = aio.secure_channel(address, credentials)
            else:
                self.channel = aio.insecure_channel(address)
            
            self.stub = streaming_pb2_grpc.StreamingServiceStub(self.channel)
            
            await asyncio.wait_for(self.channel.channel_ready(), timeout=5.0)
            return True
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è: {e}")
            return False
    
    def get_contract_tests(self) -> List[ContractTest]:
        """
        –ü–æ–ª—É—á–µ–Ω–∏–µ —Å–ø–∏—Å–∫–∞ –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–µ—Å—Ç–æ–≤ –∏–∑ —Ç–∞–±–ª–∏—Ü
        
        Returns:
            –°–ø–∏—Å–æ–∫ –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–µ—Å—Ç–æ–≤
        """
        tests = []
        
        # –¢–µ—Å—Ç—ã –¥–ª—è StreamAudio (–∏–∑ –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–∞–±–ª–∏—Ü—ã)
        tests.append(ContractTest(
            name="StreamAudio - valid request",
            method="StreamAudio",
            input_data={
                "prompt": "test",
                "hardware_id": "test_hardware_id"
            },
            expected_output={"type": "streaming", "has_text": True, "has_audio": True},
            expected_latency_ms=600.0,  # p95 ‚â§ 600ms
            description="–£—Å–ø–µ—à–Ω—ã–π —Å—Ü–µ–Ω–∞—Ä–∏–π StreamAudio"
        ))
        
        tests.append(ContractTest(
            name="StreamAudio - empty prompt",
            method="StreamAudio",
            input_data={
                "prompt": "",
                "hardware_id": "test_hardware_id"
            },
            expected_error_code="INVALID_ARGUMENT",
            expected_latency_ms=100.0,  # < 100ms
            should_fail=True,
            description="–í–∞–ª–∏–¥–∞—Ü–∏—è: –ø—É—Å—Ç–æ–π prompt"
        ))
        
        tests.append(ContractTest(
            name="StreamAudio - empty hardware_id",
            method="StreamAudio",
            input_data={
                "prompt": "test",
                "hardware_id": ""
            },
            expected_error_code="INVALID_ARGUMENT",
            expected_latency_ms=100.0,  # < 100ms
            should_fail=True,
            description="–í–∞–ª–∏–¥–∞—Ü–∏—è: –ø—É—Å—Ç–æ–π hardware_id"
        ))
        
        tests.append(ContractTest(
            name="StreamAudio - with screenshot",
            method="StreamAudio",
            input_data={
                "prompt": "test",
                "hardware_id": "test_hardware_id",
                "screenshot": "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg=="  # 1x1 PNG
            },
            expected_output={"type": "streaming", "has_text": True},
            expected_latency_ms=800.0,  # p95 ‚â§ 800ms
            description="StreamAudio —Å –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ–º"
        ))
        
        # –¢–µ—Å—Ç—ã –¥–ª—è InterruptSession (–∏–∑ –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–∞–±–ª–∏—Ü—ã)
        tests.append(ContractTest(
            name="InterruptSession - valid request",
            method="InterruptSession",
            input_data={
                "hardware_id": "test_hardware_id"
            },
            expected_output={"success": True},
            expected_latency_ms=100.0,  # < 100ms
            description="–£—Å–ø–µ—à–Ω–æ–µ –ø—Ä–µ—Ä—ã–≤–∞–Ω–∏–µ —Å–µ—Å—Å–∏–∏"
        ))
        
        tests.append(ContractTest(
            name="InterruptSession - empty hardware_id",
            method="InterruptSession",
            input_data={
                "hardware_id": ""
            },
            expected_error_code="INVALID_ARGUMENT",
            expected_latency_ms=50.0,  # < 50ms
            should_fail=True,
            description="–í–∞–ª–∏–¥–∞—Ü–∏—è: –ø—É—Å—Ç–æ–π hardware_id"
        ))
        
        tests.append(ContractTest(
            name="InterruptSession - invalid hardware_id",
            method="InterruptSession",
            input_data={
                "hardware_id": "invalid_hardware_id"
            },
            expected_output={"success": False, "message": "No active sessions found"},
            expected_latency_ms=100.0,  # < 100ms
            description="–ù–µ—Ç –∞–∫—Ç–∏–≤–Ω—ã—Ö —Å–µ—Å—Å–∏–π"
        ))
        
        # –¢–µ—Å—Ç—ã –Ω–∞ —Ç–∞–π–º–∞—É—Ç—ã (–¥–æ–±–∞–≤–ª—è–µ–º –¥–ª–∏—Ç–µ–ª—å–Ω—ã–π –∑–∞–ø—Ä–æ—Å)
        tests.append(ContractTest(
            name="StreamAudio - timeout",
            method="StreamAudio",
            input_data={
                "prompt": "test",
                "hardware_id": "test_hardware_id"
            },
            expected_error_code="DEADLINE_EXCEEDED",
            expected_latency_ms=10000.0,  # –¢–∞–π–º–∞—É—Ç 10s
            should_fail=True,
            description="–¢–∞–π–º–∞—É—Ç –∑–∞–ø—Ä–æ—Å–∞ (10 —Å–µ–∫—É–Ω–¥)"
        ))
        
        return tests
    
    async def run_test(self, test: ContractTest) -> tuple[TestResult, str]:
        """
        –ó–∞–ø—É—Å–∫ –æ–¥–Ω–æ–≥–æ –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–µ—Å—Ç–∞
        
        Args:
            test: –ö–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–µ—Å—Ç
        
        Returns:
            (—Ä–µ–∑—É–ª—å—Ç–∞—Ç, —Å–æ–æ–±—â–µ–Ω–∏–µ)
        """
        try:
            start_time = asyncio.get_event_loop().time()
            
            if test.method == "StreamAudio":
                # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å
                request = streaming_pb2.StreamRequest(
                    prompt=test.input_data.get("prompt", ""),
                    hardware_id=test.input_data.get("hardware_id", ""),
                    screenshot=test.input_data.get("screenshot") or None
                )
                
                # –í—ã–∑—ã–≤–∞–µ–º RPC —Å —Ç–∞–π–º–∞—É—Ç–æ–º
                timeout = 10.0 if test.name == "StreamAudio - timeout" else 5.0
                
                try:
                    response_count = 0
                    has_text = False
                    has_audio = False
                    
                    async for response in self.stub.StreamAudio(request, timeout=timeout):
                        response_count += 1
                        content_type = response.WhichOneof("content")
                        
                        if content_type == "text_chunk":
                            has_text = True
                        elif content_type == "audio_chunk":
                            has_audio = True
                        elif content_type == "error_message":
                            # –ü—Ä–æ–≤–µ—Ä—è–µ–º –∫–æ–¥ –æ—à–∏–±–∫–∏
                            if test.expected_error_code:
                                return (TestResult.PASS, f"–ü–æ–ª—É—á–µ–Ω–∞ –æ–∂–∏–¥–∞–µ–º–∞—è –æ—à–∏–±–∫–∞: {response.error_message}")
                            else:
                                return (TestResult.FAIL, f"–ù–µ–æ–∂–∏–¥–∞–Ω–Ω–∞—è –æ—à–∏–±–∫–∞: {response.error_message}")
                        
                        # –û–≥—Ä–∞–Ω–∏—á–∏–≤–∞–µ–º –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –æ—Ç–≤–µ—Ç–æ–≤ –¥–ª—è smoke-—Ç–µ—Å—Ç–∞
                        if response_count >= 3:
                            break
                    
                    duration_ms = (asyncio.get_event_loop().time() - start_time) * 1000
                    
                    # –ü—Ä–æ–≤–µ—Ä—è–µ–º –æ–∂–∏–¥–∞–µ–º—ã–π —Ä–µ–∑—É–ª—å—Ç–∞—Ç
                    if test.expected_output:
                        if test.expected_output.get("has_text") and not has_text:
                            return (TestResult.FAIL, "–û–∂–∏–¥–∞–ª—Å—è text_chunk, –Ω–æ –Ω–µ –ø–æ–ª—É—á–µ–Ω")
                        if test.expected_output.get("has_audio") and not has_audio:
                            return (TestResult.FAIL, "–û–∂–∏–¥–∞–ª—Å—è audio_chunk, –Ω–æ –Ω–µ –ø–æ–ª—É—á–µ–Ω")
                    
                    # –ü—Ä–æ–≤–µ—Ä—è–µ–º latency
                    if test.expected_latency_ms and duration_ms > test.expected_latency_ms:
                        return (TestResult.FAIL, f"Latency –ø—Ä–µ–≤—ã—à–µ–Ω–∞: {duration_ms}ms > {test.expected_latency_ms}ms")
                    
                    if test.should_fail:
                        return (TestResult.FAIL, "–¢–µ—Å—Ç –¥–æ–ª–∂–µ–Ω –±—ã–ª –ø—Ä–æ–≤–∞–ª–∏—Ç—å—Å—è, –Ω–æ –ø—Ä–æ—à—ë–ª")
                    
                    return (TestResult.PASS, f"–¢–µ—Å—Ç –ø—Ä–æ–π–¥–µ–Ω (duration: {duration_ms:.2f}ms)")
                    
                except grpc.RpcError as e:
                    duration_ms = (asyncio.get_event_loop().time() - start_time) * 1000
                    error_code = e.code().name if hasattr(e.code(), 'name') else str(e.code())
                    
                    if test.expected_error_code and error_code == test.expected_error_code:
                        return (TestResult.PASS, f"–ü–æ–ª—É—á–µ–Ω–∞ –æ–∂–∏–¥–∞–µ–º–∞—è –æ—à–∏–±–∫–∞: {error_code}")
                    elif test.should_fail:
                        return (TestResult.PASS, f"–û–∂–∏–¥–∞–µ–º–∞—è –æ—à–∏–±–∫–∞ –ø–æ–ª—É—á–µ–Ω–∞: {error_code}")
                    else:
                        return (TestResult.FAIL, f"–ù–µ–æ–∂–∏–¥–∞–Ω–Ω–∞—è –æ—à–∏–±–∫–∞: {error_code} - {e.details()}")
            
            elif test.method == "InterruptSession":
                # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å
                request = streaming_pb2.InterruptRequest(
                    hardware_id=test.input_data.get("hardware_id", "")
                )
                
                try:
                    response = await self.stub.InterruptSession(request, timeout=5.0)
                    duration_ms = (asyncio.get_event_loop().time() - start_time) * 1000
                    
                    # –ü—Ä–æ–≤–µ—Ä—è–µ–º –æ–∂–∏–¥–∞–µ–º—ã–π —Ä–µ–∑—É–ª—å—Ç–∞—Ç
                    if test.expected_output:
                        if test.expected_output.get("success") is not None:
                            if response.success != test.expected_output["success"]:
                                return (TestResult.FAIL, f"–£—Å–ø–µ—à–Ω–æ—Å—Ç—å –Ω–µ —Å–æ–≤–ø–∞–¥–∞–µ—Ç: –æ–∂–∏–¥–∞–ª–æ—Å—å {test.expected_output['success']}, –ø–æ–ª—É—á–µ–Ω–æ {response.success}")
                    
                    # –ü—Ä–æ–≤–µ—Ä—è–µ–º latency
                    if test.expected_latency_ms and duration_ms > test.expected_latency_ms:
                        return (TestResult.FAIL, f"Latency –ø—Ä–µ–≤—ã—à–µ–Ω–∞: {duration_ms}ms > {test.expected_latency_ms}ms")
                    
                    if test.should_fail:
                        return (TestResult.FAIL, "–¢–µ—Å—Ç –¥–æ–ª–∂–µ–Ω –±—ã–ª –ø—Ä–æ–≤–∞–ª–∏—Ç—å—Å—è, –Ω–æ –ø—Ä–æ—à—ë–ª")
                    
                    return (TestResult.PASS, f"–¢–µ—Å—Ç –ø—Ä–æ–π–¥–µ–Ω (duration: {duration_ms:.2f}ms, success: {response.success})")
                    
                except grpc.RpcError as e:
                    duration_ms = (asyncio.get_event_loop().time() - start_time) * 1000
                    error_code = e.code().name if hasattr(e.code(), 'name') else str(e.code())
                    
                    if test.expected_error_code and error_code == test.expected_error_code:
                        return (TestResult.PASS, f"–ü–æ–ª—É—á–µ–Ω–∞ –æ–∂–∏–¥–∞–µ–º–∞—è –æ—à–∏–±–∫–∞: {error_code}")
                    elif test.should_fail:
                        return (TestResult.PASS, f"–û–∂–∏–¥–∞–µ–º–∞—è –æ—à–∏–±–∫–∞ –ø–æ–ª—É—á–µ–Ω–∞: {error_code}")
                    else:
                        return (TestResult.FAIL, f"–ù–µ–æ–∂–∏–¥–∞–Ω–Ω–∞—è –æ—à–∏–±–∫–∞: {error_code} - {e.details()}")
            
            return (TestResult.SKIP, "–¢–µ—Å—Ç –Ω–µ —Ä–µ–∞–ª–∏–∑–æ–≤–∞–Ω")
            
        except Exception as e:
            return (TestResult.FAIL, f"–ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: {e}")
    
    async def run_all_tests(self) -> bool:
        """
        –ó–∞–ø—É—Å–∫ –≤—Å–µ—Ö –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–µ—Å—Ç–æ–≤
        
        Returns:
            True –µ—Å–ª–∏ –≤—Å–µ —Ç–µ—Å—Ç—ã –ø—Ä–æ—à–ª–∏, False –∏–Ω–∞—á–µ
        """
        if not await self.connect():
            return False
        
        tests = self.get_contract_tests()
        
        print(f"üß™ –ó–∞–ø—É—Å–∫ {len(tests)} –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–µ—Å—Ç–æ–≤...")
        print()
        
        passed = 0
        failed = 0
        skipped = 0
        
        for test in tests:
            result, message = await self.run_test(test)
            self.results.append((test, result, message))
            
            status = "‚úÖ" if result == TestResult.PASS else "‚ùå" if result == TestResult.FAIL else "‚è≠Ô∏è"
            print(f"{status} {test.name}: {message}")
            
            if result == TestResult.PASS:
                passed += 1
            elif result == TestResult.FAIL:
                failed += 1
            else:
                skipped += 1
        
        # –ó–∞–∫—Ä—ã–≤–∞–µ–º –∫–∞–Ω–∞–ª
        if self.channel:
            await self.channel.close()
        
        print()
        print("=" * 60)
        print(f"–†–µ–∑—É–ª—å—Ç–∞—Ç—ã: {passed} –ø—Ä–æ–π–¥–µ–Ω–æ, {failed} –ø—Ä–æ–≤–∞–ª–µ–Ω–æ, {skipped} –ø—Ä–æ–ø—É—â–µ–Ω–æ")
        print("=" * 60)
        
        return failed == 0


async def main():
    """–û—Å–Ω–æ–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è"""
    import argparse
    
    parser = argparse.ArgumentParser(
        description="gRPC Contract Tests - –ø—Ä–æ–≤–µ—Ä–∫–∞ –∫–æ–Ω—Ç—Ä–∞–∫—Ç-—Ç–∞–±–ª–∏—Ü"
    )
    parser.add_argument(
        "host",
        nargs="?",
        default="nexy-server.canadacentral.cloudapp.azure.com",
        help="–•–æ—Å—Ç —Å–µ—Ä–≤–µ—Ä–∞ (–ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: nexy-server.canadacentral.cloudapp.azure.com)"
    )
    parser.add_argument(
        "port",
        nargs="?",
        type=int,
        default=50051,
        help="–ü–æ—Ä—Ç —Å–µ—Ä–≤–µ—Ä–∞ (–ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: 50051)"
    )
    
    args = parser.parse_args()
    
    print("=" * 60)
    print("gRPC Contract Tests - PR-6")
    print("=" * 60)
    print(f"–•–æ—Å—Ç: {args.host}")
    print(f"–ü–æ—Ä—Ç: {args.port}")
    print("=" * 60)
    print()
    
    tester = GrpcContractTester(args.host, args.port)
    success = await tester.run_all_tests()
    
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n‚ö†Ô∏è –ü—Ä–µ—Ä–≤–∞–Ω–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–µ–º")
        sys.exit(1)
    except Exception as e:
        print(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

